metronome/networking/src/io/iohk/metronome/networking/RemoteConnectionManager.scala

```diff
@@ -0,0 +1,339 @@
+      maxDelay: FiniteDuration,
+      maxRandomJitter: Option[FiniteDuration]
+  )
+
+  object RetryConfig {
+    import scala.concurrent.duration._
+    def default: RetryConfig = {
+      RetryConfig(500.milliseconds, 2, 30.seconds, maxRandomJitter = None)
+    }
+
+  }
+
+  private def retryConnection[F[_]: Timer: Concurrent, K](
+      config: RetryConfig,
+      failedConnectionRequest: OutGoingConnectionRequest[K]
+  ): F[OutGoingConnectionRequest[K]] = {
+    val updatedFailureCount =
+      failedConnectionRequest.numberOfFailures + 1
+    val exponentialBackoff =
+      math.pow(config.backOffFactor.toDouble, updatedFailureCount).toLong
+
+    val randomJitter =
+      config.maxRandomJitter.fold(FiniteDuration(0, TimeUnit.MILLISECONDS)) {
+        maxJitter =>
+          val randomJitterInMillis =
+            ThreadLocalRandom.current().nextLong(maxJitter.toMillis)
+          FiniteDuration(randomJitterInMillis, TimeUnit.MILLISECONDS)
+      }
+
+    val newDelay =
+      ((config.initialDelay * exponentialBackoff)
+        .min(config.maxDelay)) + randomJitter
+
+    Timer[F]
+      .sleep(newDelay)
+      .map(_ =>
+        failedConnectionRequest.copy(numberOfFailures = updatedFailureCount)
+      )
+  }
+
+  /** Connections are acquired in linear fashion i.e there can be at most one concurrent call to remote peer.
+    * In case of failure each connection will be retried infinite number of times with exponential backoff between
+    * each call.
+    */
+  private def acquireConnections[
+      F[_]: Concurrent: TaskLift: TaskLike: Timer,
+      K: Codec,
+      M: Codec
+  ](
+      encryptedConnectionProvider: EncryptedConnectionProvider[F, K, M],
+      connectionsToAcquire: ConcurrentQueue[F, OutGoingConnectionRequest[K]],
+      connectionsHandler: ConnectionHandler[F, K, M],
+      retryConfig: RetryConfig
+  ): F[Unit] = {
+
+    /** Observable is used here as streaming primitive as it has richer api than Iterant and have mapParallelUnorderedF
+      * combinator, which makes it possible to have multiple concurrent retry timers, which are cancelled when whole
+      * outer stream is cancelled
+      */
+    Observable
+      .repeatEvalF(connectionsToAcquire.poll)
+      .filterEvalF(request => connectionsHandler.isNewConnection(request.key))
+      .mapEvalF { connectionToAcquire =>
+        connectTo(encryptedConnectionProvider, connectionToAcquire)
+      }
+      .mapParallelUnorderedF(Integer.MAX_VALUE) {
+        case Left(failure) =>
+          //TODO add logging of failure
+          val failureToLog = failure.err
+          retryConnection(retryConfig, failure.connectionRequest).flatMap(
+            updatedRequest => connectionsToAcquire.offer(updatedRequest)
+          )
+        case Right(connection) =>
+          val newOutgoingConnections =
+            HandledConnection.outgoing(connection.encryptedConnection)
+          connectionsHandler.registerIfAbsent(newOutgoingConnections)
+
```

Comment:
```suggestion
          connectionsHandler.registerOrClose(newOutgoingConnections)
```
When I saw at `registerIfAbsent` in this context I expected it to return a `Boolean` and wondered why we're not closing the connection if it wasn't absent. Had to look at the implementation to remind myself that it handles closing on its own.